src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Telemetry/TelemetryScope.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+
+namespace Microsoft.AspNetCore.Razor.Telemetry;
+
+internal sealed class TelemetryScope : IDisposable
+{
+    public static readonly TelemetryScope Null = new();
+
+    private readonly ITelemetryReporter? _reporter;
+    private readonly string _name;
+    private readonly Severity _severity;
+    private readonly ImmutableArray<Property>.Builder _properties;
+    private readonly Stopwatch _stopwatch;
+    private bool _disposed;
+
+    private TelemetryScope()
+    {
+        // This constructor is only called to initialize the Null instance
+        // above. Rather than make _name, _properties, and _stopwatch
+        // nullable, we use a ! to initialize them to null for this case only.
+        _reporter = null;
+        _name = null!;
+        _properties = null!;
+        _stopwatch = null!;
+    }
+
+    private TelemetryScope(
+        ITelemetryReporter reporter,
+        string name,
+        Severity severity,
+        ImmutableArray<Property>.Builder properties)
+    {
+        _reporter = reporter;
+        _name = name;
+        _severity = severity;
+
+        // Note: The builder that is passed in always has its capacity set to at least
+        // 1 larger than the number of properties to allow the final "ellapsedms"
+        // property to be added in Dispose below.
+        _properties = properties;
+
+        _stopwatch = StopwatchPool.Default.Get();
+        _stopwatch.Restart();
+    }
+
+    public void Dispose()
+    {
+        if (_reporter is null || _disposed)
+        {
+            return;
+        }
+
+        _disposed = true;
+
+        _stopwatch.Stop();
+        _properties.Add(new("eventscope.ellapsedms", _stopwatch.ElapsedMilliseconds));
+
+        _reporter.ReportEvent(
+            _name,
+            _severity,
+            _properties.DrainToImmutable());
```

Comment:
This is the case I mentioned where we are seemingly building up an `ImmutableArray<T>`, iterating it and promptly discarding it.

Reply:
I get the concern here. I think I can probably make this spot better.

Reply:
Good catch on this - the latest changes should be a bit better.